### PR TITLE
Verify if nodes do support seccomp and do not reconcile in that case

### PR DESCRIPTION
**Pull request by saschagrunert**
#### What type of PR is this?

/kind design
/kind feature

#### What this PR does / why we need it:
The new `seccomp` package can be used to check if seccomp is enabled on
the host.
#### Which issue(s) this PR fixes:

Refers to #76 
#### Special notes for your reviewer:

This is a RFC to ensure that we're moving into the right direction. If we wanna use something like this directly in the operator, then we could throw an event on startup to inform the user if the node supports seccomp or not.

To be able to throw the event for the node, the operator would have to require permissions to list the nodes on the cluster. The node name could be injected via an environment variable, whereas the client could retrieve the node object.

#### Does this PR introduce a user-facing change?

```release-note
- Nodes not supporting seccomp will not reconcile profiles to disk. Additionally a warning event will be thrown for the config map.
```

### CODE REVIEW

File: internal/pkg/controllers/profile/profile.go [hunk 1]
```diff
@@ -108,23 +110,35 @@ func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error)
...
+	if !seccomp.New().IsSupported(logger) {
+		err := errors.New("profile not added")
+		reason := fmt.Sprintf(
+			"node %q does not support seccomp",
+			os.Getenv(config.NodeNameEnvKey),
+		)
+		logger.Error(err, "reason", reason)
+		r.record.Event(configMap, event.Warning(event.Reason(reason), err))
+		return reconcile.Result{}, errors.Wrap(err, reason)
+	}
```
Comment (saschagrunert) [inline internal/pkg/controllers/profile/profile.go:123]: _Following-up on discussion https://github.com/kubernetes-sigs/seccomp-operator/pull/85#pullrequestreview-463875699_

@hasheddan like this? Do you think we should put the operator in some `excluded` state to not spam the events? This way we could stop the operator reconciling if seccomp is disabled on a node…

Comment (hasheddan) [reply internal/pkg/controllers/profile/profile.go:123]: @saschagrunert definitely 👍 If we just don't return the error here then it should stop trying to reconcile. controller-runtime behaves in the following manner:
1. If an `error` is returned, requeue immediately
2. If `reconcile.Result{ Requeue: true }`, requeue immediately
3. If `reconcile.Result{ RequeueAfter: someTime }`, requeue after `someTime` has elapsed
4. If `reconcile.Result{}, nil`, do not requeue (will be requeued if a change to the object is observed, or after the usually very long reconcile timeout configured for the controller manager)

Comment (saschagrunert) [reply internal/pkg/controllers/profile/profile.go]: Awesome, thank you for the clarification! :+1:

File: internal/pkg/controllers/profile/profile.go [hunk 2]
```diff
@@ -90,13 +92,18 @@ func Setup(mgr ctrl.Manager, l logr.Logger) error {
...
+	log             logr.Logger
+	record          event.Recorder
+	stopReconciling bool
 }
 
 // Reconcile reconciles a ConfigMap representing a seccomp profile.
 func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	if r.stopReconciling {
+		return reconcile.Result{}, nil
+	}
```
Comment (saschagrunert) [inline internal/pkg/controllers/profile/profile.go:105]: @hasheddan do you think it makes sense to generally stop reconciling like this? The only way to restart the reconciling would be to restart the workload now.

Comment (hasheddan) [reply internal/pkg/controllers/profile/profile.go:105]: I don't think the benefit of skipping running the checks on the node on very infrequent reconciles is worth the added complexity of managing this / having to restart the workload

Comment (saschagrunert) [reply internal/pkg/controllers/profile/profile.go:105]: Sounds good, let me revert that change.

**Comment by saschagrunert** [comment]
/hold
If we can move forward with https://github.com/seccomp/containers-golang/pull/38, then I'll rebase my PR on top of that.

**Comment by pjbgf** [comment]
> @saschagrunert this is looking pretty good to me! I wonder if instead of eventing on the node we could just event on the profile `ConfigMap` with something like `Profile not added to node ____ because seccomp is not enabled`. It seems more likely that someone would be looking there than looking for events on the node after the operator was installed. What do you think?

I have mixed feeling on this: I love the traceability it gives us, only slightly concerned how noisy this may get in very large deployments (hundreds of nodes). However, I am happy with that approach for the time being and we revisit it once we land the CRD.

File: internal/pkg/node/seccomp/seccomp.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+//counterfeiter:generate . verifer
+type verifer interface {
+	ParseStatusFile(logr.Logger, string) (map[string]string, error)
+	Prctl(int, uintptr, uintptr, uintptr, uintptr) error
+}
+
+// Seccomp is the main structure of this package.
+type Seccomp struct {
+	verifier verifer
```
Comment (pjbgf) [inline internal/pkg/node/seccomp/seccomp.go:42]:
```suggestion
	verifier verifier
```

File: internal/pkg/node/seccomp/seccomp.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+	perrors "github.com/pkg/errors"
+	"golang.org/x/sys/unix"
+)
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+
+const statusFilePath = "/proc/self/status"
+
+//counterfeiter:generate . verifer
+type verifer interface {
```
Comment (pjbgf) [inline internal/pkg/node/seccomp/seccomp.go:35]:
```suggestion
type verifier interface {
```

File: internal/pkg/node/seccomp/seccomp.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+// ParseStatusFile is the status file parse implementation for the default
+// verifer.
+func (d *defaultVerifier) ParseStatusFile(
+	l logr.Logger, filePath string,
+) (map[string]string, error) {
+	return ParseStatusFile(l, filePath)
+}
+
+// Prctl is the Prctl implementation for the default verifer.
```
Comment (pjbgf) [inline internal/pkg/node/seccomp/seccomp.go:61]:
```suggestion
// Prctl is the Prctl implementation for the default verifier.
```

File: internal/pkg/node/seccomp/seccomp.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+func (d *defaultVerifier) Prctl(
+	option int, arg2, arg3, arg4, arg5 uintptr,
+) error {
+	return unix.Prctl(option, arg2, arg3, arg4, arg5)
+}
+
+// SetVerifier can be used to replace the internal verifier implementation.
+func (s *Seccomp) SetVerifier(v verifer) {
```
Comment (pjbgf) [inline internal/pkg/node/seccomp/seccomp.go:69]:
```suggestion
func (s *Seccomp) SetVerifier(v verifier) {
```

**Comment by pjbgf** [comment]
> /hold
> If we can move forward with [seccomp/containers-golang#38](https://github.com/seccomp/containers-golang/pull/38), then I'll rebase my PR on top of that.

@saschagrunert given the value this adds, I would not be opposed to merge this and then have a follow-up PR to use the `IsSupported` from upstream when that gets merged.

**Comment by saschagrunert** [comment]
> > /hold
> > If we can move forward with [seccomp/containers-golang#38](https://github.com/seccomp/containers-golang/pull/38), then I'll rebase my PR on top of that.
> 
> @saschagrunert given the value this adds, I would not be opposed to merge this and then have a follow-up PR to use the `IsSupported` from upstream when that gets merged.

Okay cool, let me just rebase and fix the typos :+1:

**Comment by saschagrunert** [comment]
/retitle Verify if nodes do support seccomp and do not reconcile in that case

**Pull request merged**